source/lambda/apiprocessor/redaction.py

```diff
@@ -41,40 +57,242 @@ def getDocumentRedaction(documentId, bucket):
+        return (400, "bad request, redacted item left number is invalid")
+            
+    # validate width number
+    if 'width' not in item:
+        return (400, "bad request, redacted item has no width number")
+
+    if isinstance(item['width'], float) == False:
+        return (400, "bad request, redacted item has invalid width number type")
+
+    if item['width'] < 0.0:
+        return (400, "bad request, redacted item width number is invalid")
+
+    # validate height number
+    if 'height' not in item:
+        return (400, "bad request, redacted item has no height number")
+
+    if isinstance(item['height'], float) == False:
+        return (400, "bad request, redacted item has invalid height number type")
+
+    if item['height'] < 0.0:
+        return (400, "bad request, redacted item height number is invalid")
+
+   # validate label string
+    if 'label' not in item:
+        return (400, "bad request, redacted item has no label")
+
+    if isinstance(item['label'], str) == False:
+        return (400, "bad request, redacted item label is not a string")
+
+    return (200, "ok")
+
+
 # retrieve the redacted items of this document
-def saveDocumentRedaction(documentId, bucket, body):
+def saveDocumentRedaction(documentId, bucket, table, body):
 
-    result = {}
-    result['status'] = False
+    statusCode = 400
+    result = "bad request"
+    
+    #pdb.set_trace()
     
-    # save the document redaction, if it parses correctly
+    # save the document redaction, if it parses and validates correctly
     try:
         redaction = json.loads(body)
         
-        S3Helper.writeToS3(json.dumps(redaction),
-                           bucket,
-                           'public/' + documentId + '/redaction.json')
-    
-        result['status'] = True
-    
     except Exception as e:
-        print("saveDocumentRedaction exception: " + str(e))
+        return (400, "invalid json document")
+        
+    # validate redaction
+    if len(redaction.keys()) != REDACTION_NUM_OF_KEYS:
+        return (400, "bad request, wrong number of redaction keys")
+
+    # uuid present
+    if 'uuid' not in redaction:
+        return (400, "bad request, no document id provided")
+
+    if isinstance(redaction['uuid'],str) == False:
+        return (400, "bad request, document uuid is not a string type")
+
+    # check doc exists in dynamodb
+    exists = checkDocumentExists(documentId, table)
     
-    return result
+    if exists == False:
+        return (404, "document doesn't exist")
+
+    # headers
+    if 'headers' not in redaction:
+        return (400, "bad request, no redaction headers list provided")
+
+    if isinstance(redaction['headers'],list) == False:
+        return (400, "bad request, redaction headers provided is not a list")
+
+    numOfHeaders = len(redaction['headers'])
+
+    if numOfHeaders > MAXIMUM_NUM_OF_HEADERS_FOOTERS:
+        return (400, "bad request, too many headers provided in list")
```

Comment:
Can the message state "bad request, too many headers provided in list. Maximum numbers of footers supported are {MAXIMUM_NUM_OF_HEADERS_FOOTERS}"

Reply:
done